src/box/memtx_tx.c

```diff
@@ -683,6 +683,17 @@ memtx_tx_acquire_ddl(struct txn *tx)
 	(void) txn_can_yield(tx, false);
 }
 
+/**
+ * Nop and remote transactions cannot possibly conflict each other by
+ * definition, so we can filter them out when aborting transactions for DDL.
+ */
+static bool
+memtx_tx_filter_nop_and_remote_txs(struct txn *txn1, struct txn *txn2)
```

Comment:
Actually, `nop` shouldn't conflict anything since it means no-op. But our `nop` is not no-op - it can be a temporary statement, and it's a bit confusing here. Let's, at least, call this function like `memtx_tx_filter_temporary_and_remote_txs` with elaboration in commentary so that it's clear that `nop` is not always a no-op.

Reply:
Also, cannot we use `txn_is_fully_local()` here instead of `txn_is_nop()`? Any temporary transaction is fully local, and it seems that any local transaction cannot conflict with purely remote one.

Reply:
> Any temporary transaction is fully local, and it seems that any local transaction cannot conflict with purely remote one.

A transaction on data-temporary spaces (for which the space metadata is replicated) can be conflicting, so we need `txn_is_nop` here.

Reply:
Oh, thanks, I see. But `txn_is_nop` is not sufficient then. If a transaction did a replace in a data-temporary space, it is `nop` but it actually can conflict with remote one since it can do DDL on data-temporary spaces.

Reply:
No, AFAIC, it is enough. A transaction contaning DDL on data-temporary spaces would have `n_new_rows > 0`, since the DDL statement would get written to the WAL (see `txn_is_fully_local`).

Reply:
AFAICS, we [update](https://github.com/tarantool/tarantool/blob/e01c20edc31b65e020ac57e3f62a7427b6e27d53/src/box/txn.c#L650) row counts only for non-data-temporary statements.

Reply:
> So, we can have local DML transaction on data-temporary space, which is nop, and we can have transaction in applier doing DDL on this data-temporary space.

If we have a transaction in the applier doing DDL, it acquires DDL, so it is forbideen to yield — even with MVCC enabled:
https://github.com/tarantool/tarantool/blob/e01c20edc31b65e020ac57e3f62a7427b6e27d53/src/box/memtx_tx.c#L679-L684

Unless it is a space format change or a new index creation, for which yielding is safe by design.

Reply:
Consider the scenario:
```
TX1: BEGIN
TX1: DML on data-temporary space
APPLIER: DDL on data-temporary space
TX1: COMMIT
```
Here applier should abort TX1 since they are operating the same space, but it doesn't work in this way in your patch because TX1 is actually `nop` (it has no rows) and APPLIER is `fully_remote`.

Reply:
You are missing the point: the applier transaction cannot yield here, otherwise it will be aborted by yield.

Reply:
Sorry for my misconception, and thanks for the catch, fixed: added another flag to track data-temporary statements, and added a test case for this scenario.

Reply:
Resolving since this is approved anyway.